File: 2022/python/day5.py
from itertools import takewhile
from copy import deepcopy

with open('../inputs/input5') as f:
	lines = f.readlines()
	crates_count = len(lines[0]) // 4

	stacks = [[] for _ in range(crates_count)]
	
	crate_lines = list(takewhile(lambda line: line != '\n', lines))[:-1]
	for crate_line in crate_lines:
		for i in range(crates_count):
			token = crate_line[i*4:(i+1)*4]
			if token[0] == '[':
				stacks[i].insert(0, token[1])

	stacks2 = deepcopy(stacks)			
	instruction_lines = [line.strip() for line in lines if line.startswith('move')]				
	for instruction in instruction_lines:
		_, m, _, f, _, t = instruction.split(' ')
		idx = len(stacks[int(t)-1])
		for i in range(int(m)):
			stacks[int(t)-1].append(stacks[int(f)-1].pop())

			stacks2[int(t)-1].insert(idx, stacks2[int(f)-1].pop())

	for stack in stacks:
		print(stack[-1], end='')
	print()
	for stack in stacks2:
		print(stack[-1], end='')
	print()

File: 2022/rs/src/day5.rs
use anyhow::Result;

// TODO: this is not very rusty :(
fn main() -> Result<()> {
    let content = include_str!("../../inputs/input5");
    let lines: Vec<_> = content.lines().collect();
    let crates_count = (lines[0].len() + 2) / 4;

    let mut stacks: Vec<Vec<char>> = vec![vec![]; 9];

    lines
        .iter()
        .take_while(|line| !line.is_empty())
        .for_each(|line| {
            for i in 0..crates_count {
                let token: String = line[i * 4..((i + 1) * 4) - 1].into();
                if token.starts_with('[') {
                    stacks[i].insert(0, token.chars().nth(1).unwrap_or_default());
                }
            }
        });

    let mut instructions: Vec<(usize, usize, usize)> = vec![];
    lines
        .iter()
        .filter(|line| line.starts_with("move"))
        .for_each(|line| {
            let tokens: Vec<_> = line.split(' ').collect();
            instructions.push((
                tokens[1].parse().unwrap_or_default(),
                tokens[3].parse().unwrap_or_default(),
                tokens[5].parse().unwrap_or_default(),
            ));
        });

    let mut stacks2 = stacks.clone();
    for (m, f, t) in instructions {
        let idx = stacks[t - 1].len();
        for _ in 0..m {
            let value = stacks[f - 1].pop().unwrap_or_default();
            stacks[t - 1].push(value);

            let value2 = stacks2[f - 1].pop().unwrap_or_default();
            stacks2[t - 1].insert(idx, value2);
        }
    }

    print!("Part 1: ");
    for stack in stacks {
        print!("{}", stack.last().unwrap_or(&'\0'));
    }
    println!();
    print!("Part 2: ");
    for stack in stacks2 {
        print!("{}", stack.last().unwrap_or(&'\0'));
    }

    Ok(())
}
